File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/PermissionPolicySetWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import org.xmtp.android.library.libxmtp.PermissionOption
import org.xmtp.android.library.libxmtp.PermissionPolicySet

class PermissionPolicySetWrapper {

    companion object {
        private fun fromPermissionOption(permissionOption: PermissionOption): String {
            return when (permissionOption) {
                PermissionOption.Allow -> "allow"
                PermissionOption.Deny -> "deny"
                PermissionOption.Admin -> "admin"
                PermissionOption.SuperAdmin -> "superAdmin"
                PermissionOption.Unknown -> "unknown"
            }
        }

        private fun createPermissionOptionFromString(permissionOptionString: String): PermissionOption {
            return when (permissionOptionString) {
                "allow" -> PermissionOption.Allow
                "deny" -> PermissionOption.Deny
                "admin" -> PermissionOption.Admin
                "superAdmin" -> PermissionOption.SuperAdmin
                else -> PermissionOption.Unknown
            }
        }
        private fun encodeToObj(policySet: PermissionPolicySet): Map<String, Any> {
            return mapOf(
                "addMemberPolicy" to fromPermissionOption(policySet.addMemberPolicy),
                "removeMemberPolicy" to fromPermissionOption(policySet.removeMemberPolicy),
                "addAdminPolicy" to fromPermissionOption(policySet.addAdminPolicy),
                "removeAdminPolicy" to fromPermissionOption(policySet.removeAdminPolicy),
                "updateGroupNamePolicy" to fromPermissionOption(policySet.updateGroupNamePolicy),
                "updateGroupDescriptionPolicy" to fromPermissionOption(policySet.updateGroupDescriptionPolicy),
                "updateGroupImagePolicy" to fromPermissionOption(policySet.updateGroupImagePolicy),
                "updateGroupPinnedFrameUrlPolicy" to fromPermissionOption(policySet.updateGroupPinnedFrameUrlPolicy),
            )
        }

        fun createPermissionPolicySetFromJson(permissionPolicySetJson: String): PermissionPolicySet {
            val jsonObj = JsonParser.parseString(permissionPolicySetJson).asJsonObject
            return PermissionPolicySet(
                addMemberPolicy = createPermissionOptionFromString(jsonObj.get("addMemberPolicy").asString),
                removeMemberPolicy = createPermissionOptionFromString(jsonObj.get("removeMemberPolicy").asString),
                addAdminPolicy = createPermissionOptionFromString(jsonObj.get("addAdminPolicy").asString),
                removeAdminPolicy = createPermissionOptionFromString(jsonObj.get("removeAdminPolicy").asString),
                updateGroupNamePolicy = createPermissionOptionFromString(jsonObj.get("updateGroupNamePolicy").asString),
                updateGroupDescriptionPolicy = createPermissionOptionFromString(jsonObj.get("updateGroupDescriptionPolicy").asString),
                updateGroupImagePolicy = createPermissionOptionFromString(jsonObj.get("updateGroupImagePolicy").asString),
                updateGroupPinnedFrameUrlPolicy = createPermissionOptionFromString(jsonObj.get("updateGroupPinnedFrameUrlPolicy").asString)
            )
        }

        fun encodeToJsonString(policySet: PermissionPolicySet): String {
            val gson = GsonBuilder().create()
            val obj = encodeToObj(policySet)
            return gson.toJson(obj)
        }
    }
}

File: ios/Wrappers/PermissionPolicySetWrapper.swift
//
//  PermissionPolicySetWrapper.swift
//  XMTPReactNative
//
//  Created by Naomi Plasterer on 7/1/24.
//

import Foundation
import XMTP

class PermissionPolicySetWrapper {
    static func fromPermissionOption(_ permissionOption: XMTP.PermissionOption) -> String {
        switch permissionOption {
        case .allow:
            return "allow"
        case .deny:
            return "deny"
        case .admin:
            return "admin"
        case .superAdmin:
            return "superAdmin"
        case .unknown:
            return "unknown"
        }
    }
    
    static func createPermissionOption(from string: String) -> PermissionOption {
        switch string {
        case "allow":
            return .allow
        case "deny":
            return .deny
        case "admin":
            return .admin
        case "superAdmin":
            return .superAdmin
        default:
            return .unknown
        }
    }
    
    
    static func encodeToObj(_ policySet: XMTP.PermissionPolicySet) -> [String: Any] {
        
        return [
            "addMemberPolicy": fromPermissionOption(policySet.addMemberPolicy),
            "removeMemberPolicy": fromPermissionOption(policySet.removeMemberPolicy),
            "addAdminPolicy": fromPermissionOption(policySet.addAdminPolicy),
            "removeAdminPolicy": fromPermissionOption(policySet.removeAdminPolicy),
            "updateGroupNamePolicy": fromPermissionOption(policySet.updateGroupNamePolicy),
            "updateGroupDescriptionPolicy": fromPermissionOption(policySet.updateGroupDescriptionPolicy),
            "updateGroupImagePolicy": fromPermissionOption(policySet.updateGroupImagePolicy),
            "updateMessageDisappearingPolicy": fromPermissionOption(policySet.updateMessageDisappearingPolicy)
        ]
    }
    public static func createPermissionPolicySet(from json: String) throws -> PermissionPolicySet {
        guard let data = json.data(using: .utf8) else {
            throw WrapperError.decodeError("Failed to convert PermissionPolicySet JSON string to data")
        }
        
        guard let jsonObject = try? JSONSerialization.jsonObject(with: data, options: []),
              let jsonDict = jsonObject as? [String: Any] else {
            throw WrapperError.decodeError("Failed to parse PermissionPolicySet JSON data")
        }
        
        return PermissionPolicySet(
            addMemberPolicy: createPermissionOption(from: jsonDict["addMemberPolicy"] as? String ?? ""),
            removeMemberPolicy: createPermissionOption(from: jsonDict["removeMemberPolicy"] as? String ?? ""),
            addAdminPolicy: createPermissionOption(from: jsonDict["addAdminPolicy"] as? String ?? ""),
            removeAdminPolicy: createPermissionOption(from: jsonDict["removeAdminPolicy"] as? String ?? ""),
            updateGroupNamePolicy: createPermissionOption(from: jsonDict["updateGroupNamePolicy"] as? String ?? ""),
            updateGroupDescriptionPolicy: createPermissionOption(from: jsonDict["updateGroupDescriptionPolicy"] as? String ?? ""),
            updateGroupImagePolicy: createPermissionOption(from: jsonDict["updateGroupImagePolicy"] as? String ?? ""),
            updateMessageDisappearingPolicy: createPermissionOption(from: jsonDict["updateMessageDisappearingPolicy"] as? String ?? "")
        )
    }
    
    static func encodeToJsonString(_ policySet: XMTP.PermissionPolicySet) throws -> String {
        let obj = encodeToObj(policySet)
        let data = try JSONSerialization.data(withJSONObject: obj)
        guard let result = String(data: data, encoding: .utf8) else {
            throw WrapperError.encodeError("could not encode permission policy")
        }
        return result
    }
    
}
